File: providers/aws/src/main/java/cloudspec/aws/ec2/resource/nested/EC2EbsBlockDevice.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.resource.nested;

import cloudspec.annotation.AssociationDefinition;
import cloudspec.annotation.PropertyDefinition;
import cloudspec.aws.ec2.resource.EC2SnapshotResource;
import software.amazon.awssdk.services.ec2.model.EbsBlockDevice;

import java.util.Objects;
import java.util.Optional;

public class EC2EbsBlockDevice {
    @PropertyDefinition(
            name = "delete_on_termination",
            description = "Indicates whether the EBS volume is deleted on instance termination"
    )
    private final Boolean deleteOnTermination;

    @PropertyDefinition(
            name = "iops",
            description = "The number of I/O operations per second (IOPS) that the volume supports"
    )
    private final Integer iops;

    @AssociationDefinition(
            name = "snapshot",
            description = "The ID of the snapshot",
            targetClass = EC2SnapshotResource.class
    )
    private final String snapshotId;

    @PropertyDefinition(
            name = "volume_size",
            description = "The size of the volume, in GiB"
    )
    private final Integer volumeSize;

    @PropertyDefinition(
            name = "volume_type",
            description = "The volume type",
            exampleValues = "gp2, st1, sc1, standard"
    )
    private final String volumeType;

    @PropertyDefinition(
            name = "encrypted",
            description = "Indicates whether the encryption state of an EBS volume is changed while being restored from a backing snapshot"
    )
    private final Boolean encrypted;

    public EC2EbsBlockDevice(Boolean deleteOnTermination, Integer iops, String snapshotId, Integer volumeSize,
                             String volumeType, Boolean encrypted) {
        this.deleteOnTermination = deleteOnTermination;
        this.iops = iops;
        this.snapshotId = snapshotId;
        this.volumeSize = volumeSize;
        this.volumeType = volumeType;
        this.encrypted = encrypted;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (o == null || (getClass() != o.getClass() && !(o instanceof EbsBlockDevice))) {
            return false;
        }

        if (o instanceof EbsBlockDevice) {
            return sdkEquals((EbsBlockDevice) o);
        }

        EC2EbsBlockDevice that = (EC2EbsBlockDevice) o;
        return Objects.equals(deleteOnTermination, that.deleteOnTermination) &&
                Objects.equals(iops, that.iops) &&
                Objects.equals(snapshotId, that.snapshotId) &&
                Objects.equals(volumeSize, that.volumeSize) &&
                Objects.equals(volumeType, that.volumeType) &&
                Objects.equals(encrypted, that.encrypted);
    }

    private boolean sdkEquals(EbsBlockDevice that) {
        return Objects.equals(deleteOnTermination, that.deleteOnTermination()) &&
                Objects.equals(iops, that.iops()) &&
                Objects.equals(snapshotId, that.snapshotId()) &&
                Objects.equals(volumeSize, that.volumeSize()) &&
                Objects.equals(volumeType, that.volumeTypeAsString()) &&
                Objects.equals(encrypted, that.encrypted());
    }

    @Override
    public int hashCode() {
        return Objects.hash(deleteOnTermination, iops, snapshotId, volumeSize, volumeType, encrypted);
    }

    public static EC2EbsBlockDevice fromSdk(EbsBlockDevice ebsBlockDevice) {
        return Optional.ofNullable(ebsBlockDevice)
                       .map(v -> new EC2EbsBlockDevice(
                               v.deleteOnTermination(),
                               v.iops(),
                               v.snapshotId(),
                               v.volumeSize(),
                               v.volumeTypeAsString(),
                               v.encrypted()
                       ))
                       .orElse(null);
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/nested/EC2EbsBlockDevice.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2.nested

import cloudspec.annotation.AssociationDefinition
import cloudspec.annotation.PropertyDefinition
import cloudspec.aws.ec2.EC2Snapshot
import software.amazon.awssdk.services.ec2.model.VolumeType

data class EC2EbsBlockDevice(
        @property:PropertyDefinition(
                name = PROP_DELETE_ON_TERMINATION,
                description = PROP_DELETE_ON_TERMINATION_D
        )
        val deleteOnTermination: Boolean?,

        @property:PropertyDefinition(
                name = PROP_IOPS,
                description = PROP_IOPS_D
        )
        val iops: Int?,

        @property:AssociationDefinition(
                name = ASSOC_SNAPSHOT,
                description = ASSOC_SNAPSHOT_D,
                targetClass = EC2Snapshot::class
        )
        val snapshotId: String?,

        @property:PropertyDefinition(
                name = PROP_VOLUME_SIZE,
                description = PROP_VOLUME_SIZE_D
        )
        val volumeSize: Int?,

        @property:PropertyDefinition(
                name = PROP_VOLUME_TYPE,
                description = PROP_VOLUME_TYPE_D
        )
        val volumeType: VolumeType?,

        @property:PropertyDefinition(
                name = PROP_ENCRYPTED,
                description = PROP_ENCRYPTED_D
        )
        val encrypted: Boolean?
) {
    companion object {
        const val PROP_DELETE_ON_TERMINATION = "delete_on_termination"
        const val PROP_DELETE_ON_TERMINATION_D = "Indicates whether the EBS volume is deleted on instance termination"
        const val PROP_IOPS = "iops"
        const val PROP_IOPS_D = "The number of I/O operations per second (IOPS) that the volume supports"
        const val ASSOC_SNAPSHOT = "snapshot"
        const val ASSOC_SNAPSHOT_D = "The ID of the snapshot"
        const val PROP_VOLUME_SIZE = "volume_size"
        const val PROP_VOLUME_SIZE_D = "The size of the volume, in GiB"
        const val PROP_VOLUME_TYPE = "volume_type"
        const val PROP_VOLUME_TYPE_D = "The volume type"
        const val PROP_ENCRYPTED = "encrypted"
        const val PROP_ENCRYPTED_D = "Indicates whether the encryption state of an EBS volume is changed while being restored from a backing snapshot"
    }
}
